### PR TITLE
Add symfony event for public links shared by email

### PR DESCRIPTION
**Pull request by sharidas**
Adding symfony event for public links shared by email.
When an email is triggered, the recipient, the cc or
bcc and the link data is shared using the event.

Signed-off-by: Sujith H <sharidasan@owncloud.com>

## Description

Add new symfony event, when a public link is shared by email. The details include:
a) recipient
b) link
d) cc or bcc

## Motivation and Context

Add new symfony event, when a public link is shared by email. The details include:
a) recipient
b) link
d) cc or bcc

## How Has This Been Tested?

- [x] Tested if the event is called using debugger
- [x] With the unit test.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: lib/private/Share/MailNotifications.php [hunk 1]
```diff
@@ -186,6 +192,8 @@ public function sendLinkShareMail($recipient, $filename, $link, $expiration, $pe
 		$subject = (string)$this->l->t('%s shared »%s« with you', [$this->senderDisplayName, $filename]);
 		list($htmlBody, $textBody) = $this->createMailBody($filename, $link, $expiration, $personalNote);
 
+		$event = new GenericEvent(null, ['link' => $link, 'recipient' => $recipient, 'ccorbcc' => $options]);
```
Comment (PVince81) [inline lib/private/Share/MailNotifications.php]: did you test with multiple recipients ?

Comment (PVince81) [reply lib/private/Share/MailNotifications.php]: are we getting a single event with multiple recipients or multiple events ?

in the former case the attribute should then be called `recipients`, plural

Comment (sharidas) [reply lib/private/Share/MailNotifications.php]: A single event with multiple recipients. The `recipient` here is the `to` in the mail. The manual test in the UI makes multiple bcc recipients. And that gets populated in `$options`

Comment (PVince81) [reply lib/private/Share/MailNotifications.php]: Passing in the options array as is as a different attribute name is confusing. What happens if in the future we add more options that are unrelated to cc/bcc ?

I suggest we split semantically into three attributes: "to", "cc" and "bcc" so it's very clear to the event consumer that what is coming in are these fields.

Comment (sharidas) [reply lib/private/Share/MailNotifications.php]: Splitted the attributes to 'to', 'cc' and 'bcc'.

File: lib/private/Share/MailNotifications.php [hunk 2]
```diff
@@ -186,6 +192,17 @@ public function sendLinkShareMail($recipient, $filename, $link, $expiration, $pe
...
 		list($htmlBody, $textBody) = $this->createMailBody($filename, $link, $expiration, $personalNote);
 
+		/**
+		 * The event consumer of share.sendmail would have following data
+		 * - link ( the public link created )
+		 * - to ( the to recipient in mail )
+		 * - cc ( the cc recipient in mail )
+		 * - bcc ( the bcc recipient in mail )
+		 */
+		$ccRecipients = isset($options['cc']) === true ? $options['cc'] : '';
```
Comment (butonic) [inline lib/private/Share/MailNotifications.php]: `$ccRecipients = isset($options['cc']) ? $options['cc'] : '';`
`$bccRecipients = isset($options['bcc']) ? $options['bcc'] : '';`

File: tests/lib/Share/MailNotificationsTest.php [hunk 3]
```diff
@@ -29,10 +29,13 @@
 use OCP\IUser;
 use OCP\Mail\IMailer;
 use OCP\Util;
+use Symfony\Component\EventDispatcher\EventDispatcher;
+use Symfony\Component\EventDispatcher\GenericEvent;
 use Test\TestCase;
 
 /**
  * Class MailNotificationsTest
+ * @group DB
```
Comment (butonic) [inline tests/lib/Share/MailNotificationsTest.php]: why does this test now need db access?

**Review by butonic — CHANGES_REQUESTED** [review]

File: tests/lib/Share/MailNotificationsTest.php (hunk 3 above)
Comment (sharidas) [reply tests/lib/Share/MailNotificationsTest.php]: I have updated the code and brought back `@group DB`. Basically trying to improve code coverage.

**Comment by PVince81** [comment]
ignoring untestable codecov stuff, merging

**Comment by PVince81** [comment]
@sharidas please backport

**Comment by sharidas** [comment]
Backport PR: https://github.com/owncloud/core/pull/31632

**Pull request merged**